File: leetcode/1372. Longest ZigZag Path in a Binary Tree/solution.rb
def longest_zig_zag(root)
  left, right = ["left", "right"]
  ret = 0
  def visit(dir, node, step)
    return if node.nil?
    ret = [step, ret].max

    if dir == left
      visit(left, node.left, step + 1)
      visit(right, node.right, 1)
    else
      visit(left, node.right, step + 1)
      visit(right, node.left, 1)
    end
  end

  visit(left, root, 0), visit(right, root, 0)
  ret
end

File: leetcode/1372. Longest ZigZag Path in a Binary Tree/solution.py
class Solution:
    def __init__(self):
        self.ret = 0

    def dfs(self, flag, node, step):
        if not node: return 0
        self.ret = max(self.ret, step)

        if flag == "left":
            self.dfs("right", node.left, step + 1)
            self.dfs("left", node.left, 0)
        else:
            self.dfs("left", node.right, step + 1)
            self.dfs("right", node.right, 0)

    def longestZigZag(self, root):
        self.dfs("left", root, 0)
        self.dfs("right", root, 0)
        return self.ret

